### PR TITLE
Fedora compat fixes for package version checks and dnf bug workaround

### PR DESCRIPTION
**Pull request by maxamillion**
Fix a couple things that don't work with Fedora:
- Don't attempt to version check components, they aren't going to match the hard coded RHEL package versions.
- Work around a [bug in dnf](https://bugzilla.redhat.com/show_bug.cgi?id=1199432)'s python API that throws an error when a package spec has a `*` character in the Release field

Once we're done with review, I'll squash commits.

### CODE REVIEW

File: playbooks/common/openshift-cluster/config.yml [hunk 1]
```diff
@@ -17,6 +17,24 @@
...
+  post_tasks:
+  - action: openshift_health_check
+    args:
+      checks:
+      - disk_availability
+      - memory_availability
+      - package_availability
+      - docker_image_availability
+      - docker_storage
+    when: ansible_distribution != "Fedora"
```
Comment (sdodson) [inline playbooks/common/openshift-cluster/config.yml]: Not checking on Fedora is probably the right technical solution but I wonder if we should just forego these checks on origin installs where we have less ability to enforce standards around the OS and config and we're more willing to accept workable but not supported configs? @rhcarvalho @sosiouxme what do you say?

**Comment by sdodson** [comment]
aos-ci-test

File: playbooks/common/openshift-cluster/config.yml (hunk 1 above)
Comment (maxamillion) [inline playbooks/common/openshift-cluster/config.yml]: @sdodson - yeah, I think that makes sense. Folks running upstream bits "hot off the press" are going to have a higher potential to have versions outside the known test matrix. :+1: 

If we're waiting for others to chime in I'll hold off on a change, otherwise I'll get a fix in. Just let me know.

**Comment by sdodson** [comment]
[merge]

**Comment by maxamillion** [comment]
@sdodson I don't think the Jenkins failures were related to these changes. Please advise.

**Comment by sdodson** [comment]
[merge]

**Comment by sdodson** [comment]
It should get another round through when it gets to the top of the merge queue.

**Comment by michaelgugino** [comment]
Please see my comment above before merging.  I believe there is a '!=' where there should be an '==' in the Fedora case.

Also, this looks like the role will execute twice against all hosts, only the post tasks are being filtered.

**Comment by sdodson** [comment]
@michaelgugino yeah you're right

**Comment by maxamillion** [comment]
thanks!

File: playbooks/common/openshift-cluster/config.yml (hunk 1 above)
Comment (sosiouxme) [inline playbooks/common/openshift-cluster/config.yml]: `openshift_health_checker` doesn't run any tasks -- none are even available; it just defines plugins. So task(s) use the plugins after the role is included.

Comment (sosiouxme) [inline playbooks/common/openshift-cluster/config.yml]: I agree that it isn't too helpful to run checks for supportable resources on Fedora. It would be nice to have the package requirements correct for Fedora but if we're not going to commit to keeping up with that (and I don't feel like we will yet) then best just to leave `package_availability` out too (I think it already exempts itself in some cases?).

I think there is some value in applying the `docker_image_availability` check on Fedora, as that will let you know if your install will end up trying to use images that don't exist, and that seems pretty relevant to me.

There's no particular need to have separate plays here. So maybe:

```
 - name: Verify Install Requirements
   hosts: oo_all_hosts
   roles:
   - openshift_health_checker
   vars:
   - r_openshift_health_checker_playbook_context: install
   post_tasks:

   - name: Verify Requirements - EL
     when: ansible_distribution != "Fedora"
     action: openshift_health_check
     args:
       checks:
       - disk_availability
       - memory_availability
       - package_availability
       - package_version
       - docker_image_availability
       - docker_storage

   - name: Verify Requirements - Fedora
     when: ansible_distribution == "Fedora"
     action: openshift_health_check
     args:
       checks:
       - docker_image_availability
```

File: playbooks/common/openshift-cluster/config.yml [hunk 2]
```diff
@@ -17,6 +17,24 @@
...
+  post_tasks:
+  - action: openshift_health_check
+    args:
+      checks:
+      - disk_availability
+      - memory_availability
+      - package_availability
+      - docker_image_availability
+      - docker_storage
+  when: ansible_distribution == "Fedora"
```
Comment (sosiouxme) [inline playbooks/common/openshift-cluster/config.yml]: Still has logic backward :)

This got attached to an outdated commit so repeating here:

I agree that it isn't too helpful to run checks for supportable resources on Fedora. It would be nice to have the package requirements correct for Fedora but if we're not going to commit to keeping up with that (and I don't feel like we will yet) then best just to leave `package_availability` out too (it already exempts itself anyway when yum is not in use).

I think there is some value in applying the `docker_image_availability` check on Fedora, as that will let you know if your install will end up trying to use images that don't exist, and that seems pretty relevant to me.

There's no particular need to have separate plays here. So maybe:

```
 - name: Verify Install Requirements
   hosts: oo_all_hosts
   roles:
   - openshift_health_checker
   vars:
   - r_openshift_health_checker_playbook_context: install
   post_tasks:

   - name: Verify Requirements - EL
     when: ansible_distribution != "Fedora"
     action: openshift_health_check
     args:
       checks:
       - disk_availability
       - memory_availability
       - package_availability
       - package_version
       - docker_image_availability
       - docker_storage

   - name: Verify Requirements - Fedora
     when: ansible_distribution == "Fedora"
     action: openshift_health_check
     args:
       checks:
       - docker_image_availability
```

Comment (maxamillion) [inline playbooks/common/openshift-cluster/config.yml]: Yeah, I missed the `package_availability` in my [add dnf support patch](https://github.com/openshift/openshift-ansible/pull/5101). I guess I could add that later but if there's not much value in it then I can just leave it. I'll make them a single play.

Comment (maxamillion) [inline playbooks/common/openshift-cluster/config.yml]: @sosiouxme Don't we still want memory and disk checks for Fedora?

Comment (sosiouxme) [inline playbooks/common/openshift-cluster/config.yml]: They're as much to let the admin know about support requirements as anything. Those with dev/test clusters mostly turn them off.

Comment (maxamillion) [inline playbooks/common/openshift-cluster/config.yml]: @sosiouxme FWIW, we're running in production on Fedora+Origin. :)

Comment (sosiouxme) [inline playbooks/common/openshift-cluster/config.yml]: Neat! Well I don't feel strongly either way. Nothing magical happens if your system doesn't meet the requirements, except that things probably won't run as well as you would like at scale, and Red Hat will consider you unsupported. Running checks and gating the install on it is a balancing act between annoying the user and informing them, so you have to consider whether the information is valuable to enough people to risk annoying the rest of them.

Comment (maxamillion) [inline playbooks/common/openshift-cluster/config.yml]: @sosiouxme fair point, I'll leave them out for Fedora.

Comment (rhcarvalho) [inline playbooks/common/openshift-cluster/config.yml]: Alternatively we could set the default required value to a low number, the benefit being preventing situations as described in this RFE https://github.com/openshift/openshift-ansible/issues/4413, that led to checking available space in some other paths that are actively used during the install.

To set the required value, one can set this ansible variable:

```yaml
vars:
    openshift_check_min_host_disk_gb: 0.1
```

File: playbooks/common/openshift-cluster/config.yml [hunk 3]
```diff
@@ -1,22 +1,32 @@
...
-      - package_availability
-      - package_version
-      - docker_image_availability
-      - docker_storage
+- name: Verify Install Requirements
+ hosts: oo_all_hosts
+ roles:
+ - openshift_health_checker
+ vars:
+ - r_openshift_health_checker_playbook_context: install
```
Comment (rhcarvalho) [inline playbooks/common/openshift-cluster/config.yml]: Suggestion:

```yaml
...
roles:
- role: openshift_health_checker
  r_openshift_health_checker_playbook_context: install
...
```

See https://github.com/openshift/openshift-ansible/pull/5136/files#r134479196

File: playbooks/common/openshift-cluster/config.yml [hunk 4]
```diff
@@ -1,22 +1,32 @@
...
-    args:
-      checks:
-      - disk_availability
-      - memory_availability
-      - package_availability
-      - package_version
-      - docker_image_availability
-      - docker_storage
+- name: Verify Install Requirements
+ hosts: oo_all_hosts
```
Comment (rhcarvalho) [inline playbooks/common/openshift-cluster/config.yml]: There is an indentation error here, needing to prepend an extra space here and in the rest of the block.

File: playbooks/common/openshift-cluster/config.yml (hunk 2 above)
Comment (maxamillion) [inline playbooks/common/openshift-cluster/config.yml]: @rhcarvalho would that be a Fedora specific check or a check for all distros?

Comment (rhcarvalho) [inline playbooks/common/openshift-cluster/config.yml]: The check itself only runs on hosts in either `masters`, `nodes` or `etcd` groups, unless you specify otherwise like here we're using a `when` clause in Ansible. There is nothing distro-specific.

With that configuration in my previous comment, the check should ensure there is at least 0.1 GB free in `/var`, 1 GB in `/usr/local/bin` and 1 GB in `/tmp` (or equivalent temp dir).

Maybe requiring 1 GB for `/var` would even be fair, to avoid surprises related to no disk space left?

Comment (rhcarvalho) [inline playbooks/common/openshift-cluster/config.yml]: Note: the values for `/usr/local/bin` and `/tmp` are the default, could also be configured if desired.

The more comprehensive configuration would look like:

```yaml
vars:
  openshift_check_min_host_disk_gb:
    "/var":
      masters: 0.1
      nodes: 1
      etcd: 2
    "/usr/local/bin":
      masters: 0.1
      nodes: 1
      etcd: 2
```

Replacing `0.1`, `1`, `2` with any desired value.

(/cc @sosiouxme, now I realize that configuring the value for the temp dir depends on the value that `tempfile.gettempdir()` evaluates to, which is not really good UX)

Comment (maxamillion) [inline playbooks/common/openshift-cluster/config.yml]: @rhcarvalho I understand there's nothing distro specific about the check, what I meant was: Is this a change we want for all distros or just Fedora? Because if it's not Fedora specific, then it's out of scope for this Pull Request.

Comment (rhcarvalho) [inline playbooks/common/openshift-cluster/config.yml]: @maxamillion, right. My proposal is specific to Fedora, as an alternative to disabling the check altogether.

File: playbooks/common/openshift-cluster/config.yml (hunk 3 above)
Comment (maxamillion) [inline playbooks/common/openshift-cluster/config.yml]: @rhcarvalho I'm not trying to sanitize or iterate on the tasks as they exists, I'm simply attempting to introduce Fedora compatibility. I think this change is outside the scope of this PR and should be done separately in the event it was deemed necessary for a revert, the Fedora compat would not be lost.

**Comment by maxamillion** [comment]
@sdodson I also don't think this round of failures is related to this change. Is Jekins ever happy? :)

**Comment by sosiouxme** [comment]
>
> configuring the value for the temp dir depends on the value that
> tempfile.gettempdir() evaluates to, which is not really good UX)
>

Yes, but failing halfway through an install because of a temp dir the user
has never heard of is not any better UX than failing up front...

**Comment by sdodson** [comment]
flake on https://github.com/openshift/origin/issues/13975

**Comment by sdodson** [comment]
I didn't realize that this was five commits. Since we need to do another round through both aos-ci-test and the merge queue can we squash this down to the minimum number of logical commits so we don't have unnecessary intermediate states.

**Comment by maxamillion** [comment]
@sdodson rebased and squashed

**Comment by sdodson** [comment]
aos-ci-test

**Comment by maxamillion** [comment]
@sdodson what happened with this?

**Comment by rhcarvalho** [comment]
@maxamillion there is a merge conflict in `playbooks/common/openshift-cluster/config.yml`. Please rebase against master to fix the conflict.

**Comment by maxamillion** [comment]
@rhcarvalho yeah, I got that. Was curious why this was never merged before the conflict occurred.

**Comment by sdodson** [comment]
Hard to tell, test logs are gone at this point. Lets rebase and try to get this merged.

**Comment by maxamillion** [comment]
@sdodson I don't think that failure is related to this PR, can we re-kick the tests?

**Comment by maxamillion** [comment]
@sdodson @michaelgugino please advise

**Comment by maxamillion** [comment]
Rosie doesn't like me anymore ;)

**Comment by sosiouxme** [comment]
Rosie is obsolete :) Now we have a new class of problems...

**Comment by michaelgugino** [comment]
Barring the need to rebase again, I think we should push this through if the merge queue fails.  It's a small, isolated patch that has been outstanding for quite some time.  This should have merged weeks ago.

**Comment by maxamillion** [comment]
@michaelgugino thanks! :+1:
